### DISCUSSION TITLE
Multi server support

### DISCUSSION THREAD

**Discussion by theothornhill** [joaotavora/eglot#1429] [Q&A] [answered]
Hello there!
I've been looking through other issues and PR's from earlier mentioning multi-server support. It doesn't seem to be a trivial thing to add to eglot as far as I can tell from your comments, but I believe it's a useful one. 

The easiest example where multi-server support is wanted I can come up with is JavaScript and ESlint, notably the `typescript-language-server` and `vscode-eslint`. Many of the important code actions during for example React development relies on the ESlint server. 

Is this something that could only be fixed by your earlier mentioned "server multiplexer program"?

**Comment by joaotavora** [comment]
> Is this something that could only be fixed by your earlier mentioned "server multiplexer program"?

Not really, it could also be fixed by something else, like overhauling `eglot.el`'s code to send two requests instead of one every time, and expect two responses.

But I think it's much cleaner to do the "server multiplexer".  One way of the other, someone has to do the work.  If you know Python, I encourage you to get started on the latter: I think it's the language to use for that development.

Where is the other issue, by the way? Can you link it here?

**Comment by theothornhill** [comment]
Just to be clear - you mean a program that would accept one request from Eglot (or other lsp clients), route that request to several servers, merging results from both and ship them back as one lsp compliant response to Eglot?

I just wonder whether this is something that is wanted from other lsp clients, since many of them already support multi server files.

You talked about it in #249, #180, #90 - where the last one seems like the most fruitful discussion.

**Comment by joaotavora** [comment]
> Just to be clear - you mean a program that would accept one request from Eglot (or other lsp clients), route that request to several servers, merging results from both and ship them back as one lsp compliant response to Eglot?

yes

> I just wonder whether this is something that is wanted from other lsp clients, since many of them already support multi server files.

Well, no idea if that is wanted by them, but I think this would be the best way to do it, because it's client-independent, and pretty clean. But I've never tried it, it's vapourware :-)

> You talked about it in #249, #180, #90 - where the last one seems like the most fruitful discussion.

Yes, I had a feeling there was an issue for that already.  Let's close this one then.

**Comment by joaotavora** [comment]
I've found some time to work on this.  So I'm reopening, converting this to a conversation and linking all (or most) other issues here.

**Reply by jsadusk** [reply]
I just noticed you're working on this. I have a few real world use cases for this and would be happy to give you examples or test. I'm currently using lsp-mode only to get this functionality and would switch back in a heartbeat if this worked.

**Reply by garyo** [reply]
Another use case: C++ with clangd and a copilot (TabbyML or others). Can't wait for this!

**Reply by anarcat** [reply]
Trying to figure out if eglot supports multiple LSPs, and I'm a little confused as I read in https://github.com/joaotavora/eglot/issues/90#issuecomment-1060042169 that "this supported by `eglot-alternatives`", yet this issue here seems to say it isn't...

My use case is to run `ruff` alongside `pylsp` to both have the linter/formatter (ruff) available while having another tool (pylsp) do the classic completion stuff...

**Reply by jsadusk** [reply]
@anarcat for what you're trying to do you don't actually need this functionality. `pylsp` can load `ruff` as a plugin, and then `eglot` can use it as a single LSP server. Alternatively, you could use `ruff` in non-lsp mode as an `flymake` checker.

My use case is that the company I work for uses a commercial linting tool called `trunk`. Its a cross language meta linter, and it only provides an LSP interface. It does not provide any completion functionality. I'm developing both python and rust, and to be able to use the company wide linting rules, I need to be able to run the `trunk` lsp server alongside either a python or rust completion lsp (I'm using `jedi-language-server` and `rust-analyzer`). 

I'm actually back using eglot because it is just better than lsp-mode, especially running over tramp. I'm running trunk as a command line utility to get results, but I really miss having inline checking. I tried using a couple lsp proxy projects, I haven't gotten any of them to work.

The use case of linters that provide an lsp interface separate from your completion server seems to be growing. This will be amazingly useful functionality when its implemented.

**Reply by anarcat** [reply]
that's really nice to hear, all this is kind of confusing to me, so i'm really glad i got this clarified...

(for what it's worth, i'm currently using lsp-mode, but that's just historical inertia on my part, i think, now that eglot is in emacs, i'll try to switch over but, alas, that's just ... work...)

**Reply by LionyxML** [reply]
> Trying to figure out if eglot supports multiple LSPs, and I'm a little confused as I read in [#90 (comment)](https://github.com/joaotavora/eglot/issues/90#issuecomment-1060042169) that "this supported by `eglot-alternatives`", yet this issue here seems to say it isn't...
> 
> My use case is to run `ruff` alongside `pylsp` to both have the linter/formatter (ruff) available while having another tool (pylsp) do the classic completion stuff...

I think it means: it supports multiple LSPs, but only ONE per buffer, so you can have typescript support in a buffer and pyright on another, but not composable (typescript + tailwind, for example) on the same buffer (as lsp-mode supports).

**Reply by haze** [reply]
@joaotavora just curious, were there any updates on this front?

**Comment by farynaio** [comment]
Typescript/ES6 + Tailwindcss, another good example

**Reply by LionyxML** [reply]
Regarding **multiple** servers per buffer, I would add some of my daily use cases:

- typescript (tsls server), eslint (as a lsp server, much faster), and tailwind
- typescript (tsls server), biome (as eslint replacement written in rust, thus faster), and tailwind
- typescript (tsls server), vls (Vue.js server), eslint/biome and tailwind
- htmx-lsp, tailwind, html (from vscode-html-language-server) and or cssls

**Reply by Askath** [reply]
Typescript tsls, Angular ls, eslint would also be the case for me.

**Comment by ed7ed** [comment]
@joaotavora Hello, are there plans to make it live in the nearest future?

**Reply by joaotavora** [reply]
No, not in the "nearest" future, no.  But it's not exactly abandoned, I just don't have time.  You can try writing one of these multiplexers yourself.  I'm trying it in C++ because reasons, but making that works on windows/mac/linux is kind of a pain. A different approach could be using Python and some ready-made JSONRPC library.  I'll not be pursuing it, though, but maybe it's doable.

**Reply by garyo** [reply]
I just wrote a very simple LSP multiplexer in python. It's only been tested _very_ lightly so far, and it's just a one-day spare-time project at the moment, but you're welcome to check it out: https://github.com/garyo/lsp-multiplexer

**Reply by ed7ed** [reply]
> No, not in the "nearest" future, no. But it's not exactly abandoned, I just don't have time. You can try writing one of these multiplexers yourself. I'm trying it in C++ because reasons, but making that works on windows/mac/linux is kind of a pain. A different approach could be using Python and some ready-made JSONRPC library. I'll not be pursuing it, though, but maybe it's doable.

Understandable, take your time.

**Reply by jdtsmith** [reply]
FYI, I just proposed such a multiplexer addition to the rust-based [`emacs-lsp-booster`](https://github.com/blahgeek/emacs-lsp-booster) (before I found this discussion); see blahgeek/emacs-lsp-booster#38.  

I think it's a natural fit there, since it already has mastered communicating with a subordinate server, caching requests to and responses from it, and live-translating to emacs-bytecode for maximum communication speed.

Obviously the "multiplexer" aspect needs a bit of thought.  It's not quite enough to have the user specify this ("this request goes to this server, these requests to that one").  I think it probably needs to spy on and get involved in the `initialize` traffic so it can check the advertised `capabilities`.  Curious how people have solved/thought to solve that aspect.

**Reply by cowboyd** [reply]
With a multiplexer, how would you handle the case where one server is requesting incremental text updates and another is requesting the full text. I.e.

server 1 capabilities
```
{
  "textDocumentSync": 1
}
```

server 2 capabilities
```
{
  "textDocumentSync": {
    "openClose": true,
    "change": 2
  }
}
```

Do we just go with lowest common denominator (full text sync)? Or what about the case where one server does not want to be notified at all?

**Reply by joaotavora** [reply]
> Do we just go with lowest common denominator (full text sync)

That works. You could also have the multiplexer keep its own view of the document, then feed each server what it supports.

> Or what about the case where one server does not want to be notified at all?

That would be bizarre (what would be the point of such a server?) but not hard to implement, just withhold the message

**Reply by MagicRB** [reply]
Different idea I had, even if `eglot` doesn't support multiple LSP servers by itself, it could delegate the heavy lifting to a separate process but also be aware of the fact its talking to a multiplexed server. Specifically I mean a simple UI for listing connected servers, perhaps restarting them, that kind of thing

**Reply by joaotavora** [reply]
> Different idea I had, even if eglot doesn't support multiple LSP servers by itself, it could delegate the heavy lifting to a separate process 

That's exactly the idea of the multiplexer server :-)

**Reply by cowboyd** [reply]
I went ahead and took a stab at this, and you can find the result here https://github.com/thefrontside/lspx

It's a standalone executable that does very simple matching of requests to the capabilities of the lsp servers it is managing, and then merges those responses. There still is a lot to do, but I am successfully running both the Deno LSP and the Tailwind LSP in emacs at the same time which is a big win.

While there are a lot of unhandled cases and bugs to be found, there is a test suite, and hooks extending how requests and notifications are matched and merged.

**Reply by merisbahti** [reply]
I’ve been trying lspx for my svelte project, but it seems to break after i add tailwind and/or vscode eslint lsp. I am an emacs noob and might be configuring eglot incorrectly, how do you configure eglot/lspx to use ts and tailwind?

**Reply by joaotavora** [reply]
> I’ve been trying lspx for my svelte project, but it seems to break after i add tailwind and/or vscode eslint lsp. I am an emacs noob and might be configuring eglot incorrectly, how do you configure eglot/lspx to use ts and tailwind?

At this point I think it'd make sense to move such questions to the lspx tracker.  Good work @cowboyd !

**Comment by cowboyd** [comment]
@merisbahti I'd love to learn how it is breaking for you, and as suggested, we should discuss that over on the lspx repo.

The information that is relevant to eglot is that you can always use the prefix argument `C-u eglot` to always have it prompt you for what command to run to start the language server, so in your case you would enter `lspx --lsp "tailwindcss-language-sever --stdio" --lsp "eslint-lsp --stdio" as the command.

For more automagical detection. I'm considering an emacs package that would extend `eglot` by changing the default lsp  for a language to be `lspx` in the event that there was a `.lspxrc` directory in the project. Does that sound like a reasonable approach @joaotavora ?

**Reply by joaotavora** [reply]
Yes, sounds reasonable. But, even better, if `lspx` ever becomes a standard tool, it'd make sense for this small Elisp util to be in Eglot itself, so that's it's easy to see, for each entry in `eglot-server-programs`, that there are some as alternatives that are multi-server. But my advice in the meantime is to make good use of `C-u M-x eglot` and its history navigation capabilities. Let's do that and wait a bit for your program to grow and mature before we commit to having `eglot.el` having some kind of multi-server support.

**Comment by KristianAN** [comment]
@cowboyd @joaotavora 
Could the work on such a multiplexer be combined somehow with the work that is done in https://github.com/blahgeek/emacs-lsp-booster ? To me it seems like there is quite a bit of overlap in what is being done. Relevant discussion here: https://github.com/blahgeek/emacs-lsp-booster/issues/38

**Reply by joaotavora** [reply]
I dunno, I guess. 

This is outside the scope of this conversation, which is not performance.

Afaik, lsp-booster does many things, one of them is transmitting JSON as Elisp byte code. Dunno if still relevant given recent parser improvements, but might be, who knows. I'd rather have Eglot keep speaking JSONRPC... Currently the transport is JSONRPC/JSON/UTF-8 but if you can make it JSONRPC/ElispBytecode and you put that change where it belongs (probably `jsonrpc.el`), then fine with me.

Other things like having the proxy/multiplexer be clever about what LSP things it delivers to the client are of course OK, as long as what Eglot sees is an LSP server. But, again, that's outside the scope of this issue.

**Reply by cowboyd** [reply]
@KristianAN I've created a discussion on the repository for this https://github.com/thefrontside/lspx/discussions/1

**Reply by jdtsmith** [reply]
I [find](https://www.reddit.com/r/emacs/comments/1jsxamc/the_new_json_parser_is_fast/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button) with Emacs 30's native JSON parser:

> The upshot is that bytecode translation definitely reduces message sizes (often by ~40%), making it faster to read in small messages, but JSON parsing is now faster than bytecode parsing (as you'd expect), making it faster to parse large messages.

The booster still does have the advantage of caching messages, so it "gets of of the way" quickly and may still improve input latency, but it's a more subtle effect.  

I had [proposed](https://github.com/blahgeek/emacs-lsp-booster/issues/38) putting multiplex capabilities into the booster.  The author is an `lsp-mode` user so nothing brewing atm.

**Comment by Askath** [comment]
Is lspx the most promising workaround/implementation right now?

Would love to try and switch to eglot again

**Reply by Askath** [reply]
> Is lspx the most promising workaround/implementation right now?
> 
> Would love to try and switch to eglot again

Tried Lspx out with Angular and seems to work well with typescript-language-server + ngserver and htmls + ngserver.
Have not tried to add the run command for lspx to eglot-server-programs tho

**Comment by pbgc** [comment]
I'm using lspx to finally use volar + tailwind lsp (tsx + tailwind ; html + tailwind, etc) and it's working pretty well!

**Reply by jsadusk** [reply]
I just tried lspx for basedpyright + ruff, and it works great for diagnostics, but it crashes if you try to use a code action.

**Comment by fjolliton** [comment]
I tried lspx to run both "ruff server" and "ty server" (for Python code, where "ty" is an experimental type checking LSP), but it appears that eglot or lspx doesn't combine the diagnostics correctly. I see either the ones from "ruff server" after saving a file for example, or I see the ones from "ty server" if I update the code (for a fraction of a second I see the ones from "ruff server" when looking at flymake-show-buffer-diagnostics). I'm not sure who is at fault. Either lspx is not combining both diagnostics correctly, or eglot is clearing one list before getting the other list of diagnostics. I don't know the LSP protocol well enough to figure the issue.

**Reply by fjolliton** [reply]
Surprisingly, I see the exact same behavior when testing lsp-mode (which implicitly use both ruff and ty). This is puzzling. I have some doubts regarding the origin of the behavior. It clearly appears as a race condition, because when I edit the source I see sometimes the diagnostics of the other LSP, but neither lsp-mode nor eglot seem to be able to combine both list. (edit: lsp-mode is not using lspx, it runs both LSP natively)

**Reply by jdtsmith** [reply]
Check `M-x eglot-events-buffer` and see what events are coming that are conflicting.

**Reply by fjolliton** [reply]
I saw that there was an issue related to this, with some technical details: https://github.com/emacs-lsp/lsp-mode/issues/2808

Looks like it is a known behavior with flymake (for both eglot and lsp-mode), and likely to be off topic here.

**Reply by Askath** [reply]
Ah cool to see an issue about that here.
Encountered the same issue and am trying to find a workaround somehow - or how to merge diagnostics.
 I wonder if there is a way to reimplement part of flymake

**Reply by jleechpe** [reply]
The issue seems to be due to #159 clearing out the entire buffer after each LSP runs diagnostics.  Ruff/Ty will provide a set of diags, then the next language server will return it's own set or nil which removes the prior set.  

If you get rid of the added `:region (cons (point-min) (point-max))` you get to see all the diagnostics, unfortunately that also means they never get cleared.  This does also mean that even if the diagnostics were not being cleared across the entire buffer but only the region to diagnose it would only ever show whichever LSP is slowest to return for a given region since the overlap would remove the faster results.

```elisp
(defun eglot--report-to-flymake (diags version)
  "Internal helper for `eglot-flymake-backend'."
  (save-restriction
    (widen)
    (if (or (null version) (= version eglot--versioned-identifier))
        (funcall eglot--current-flymake-report-fn diags
                 ;; If the buffer hasn't changed since last
                 ;; call to the report function, flymake won't
                 ;; delete old diagnostics.  Using :region
                 ;; keyword forces flymake to delete
                 ;; them (github#159).
                 :region (cons (point-min) (point-max)))
      ;; Here, we don't have anything up to date to give Flymake: we
      ;; just want to keep whatever diagnostics it has annotated in
      ;; the buffer. However, as a nice-to-have, we still want to
      ;; signal we're alive and clear a possible "Wait" state.  We
      ;; hackingly achieve this by reporting an empty list and making
      ;; sure it pertains to a 0-length region.
      (funcall eglot--current-flymake-report-fn nil
               :region (cons (point-min) (point-min)))))
  (setq eglot--diagnostics (cons diags version)))
```

**Reply by joaotavora** [reply]
I don't understand what is being proposed here, but the whole idea of the LSP multiplexer is that Eglot shouldn't even be aware it is talking to one such thing.  So it's the multiplexer's job to behave as one of the hundreds of LSP servers out there that Eglot works fine with.  If that means buffering diagnostics from two or more different sources and comibing them before sending them to the client, then that's what it needs to do!

**Reply by Lenbok** [reply]
See also https://github.com/thefrontside/lspx/issues/14 (in particular https://github.com/thefrontside/lspx/issues/14#issuecomment-3359594308 regarding adding an agent id to diagnostics in the multiplexer plus some advising of eglot to merge when it's different agents reporting the diags)

**Comment by joaotavora** [comment]
Hey, everybody.  For the last week or so I worked a bit on https://github.com/joaotavora/rassumfrassum which seems a fairly promising LSP multiplexer IMHO.   It's in Python.  Probably has lots of bugs but works decently in my Pythonesque tests. I threw away the silly defunct C++ approach I had tried before.

**Reply by LionyxML** [reply]
Awesome! Just added to my "todo list" of things to try out! Thanks João!

**Comment by joaotavora** [comment, marked as answer]
Now that there are (at least) two alternatives to this: 

https://github.com/joaotavora/rassumfrassum

and

https://github.com/thefrontside/lspx

I'm marking this discussion closed.

**Reply by joaotavora** [reply]
A 5-strong mob thinks i should reopen the discussion :laughing: ...

**Reply by nemethf** [reply]
So, what is missing from Eglot that the three available multiplexer implementations cannot theoretically provide? 

@MagicRB wrote "a simple UI for listing connected servers, perhaps restarting them, that kind of thing".  I think a multiplexer could inject its own code-action named "Multiplexer settings/status", which cloud handle "those kinds of things" without the need to modify the LSP client.

**Reply by Askath** [reply]
> So, what is missing from Eglot that the three available multiplexer implementations cannot theoretically provide? 
> 
> 
> 
> @MagicRB wrote "a simple UI for listing connected servers, perhaps restarting them, that kind of thing".  I think a multiplexer could inject its own code-action named "Multiplexer settings/status", which cloud handle "those kinds of things" without the need to modify the LSP client. 

Last time I tried it there were still issues with diagnostics and fly make which made that solution partially unusable for me. 

I ended up going back to LSP mode and I hope for an official implementation someday

**Reply by SpyrosRoum** [reply]
> I hope for an official implementation someday

Official in what way? [rassumfrassum](https://github.com/joaotavora/rassumfrassum) is from joaotavora so even if it's not baked in eglot I'd say that's pretty "official".
Did you open tickets for the issues you encountered? No matter where or from who the multiplexer is, it's bound to have bugs and kinks at the beginning that need to be ironed out.

**Reply by joaotavora** [reply]
> Official in what way? [rassumfrassum](https://github.com/joaotavora/rassumfrassum) is from joaotavora so even if it's not baked in eglot I'd say that's pretty "official".

It's as "official" as I could make it.  I was teasing it/delaying it for years and a few months ago I bit the bullet and started working.  It wasn't as hard to do as I thought, to be honest.  The main unblocker was using Python, I have to admit.

> Did you open tickets for the issues you encountered? No matter where or from who the multiplexer is, it's bound to have bugs and kinks at the beginning that need to be ironed out.

:smiley:  I second this! :100: 

@nemethf 
>  I think a multiplexer could inject its own code-action named "Multiplexer settings/status", which cloud handle "those kinds of things" without the need to modify the LSP client.

Yes, very much so, as long as these actions are language-agnostic.  BTW @nemethf , you'll also notice a very slight break from my usual policy of "Eglot is only for (official) LSP".  It now supports a non-standard protocol extension for fast diagnostic interchange where a server can update the diagnostics for a specific aspect of the document and leave other aspects untouched.   Obvisouly the "aspects" in use by Rassumfrassum are the underlying multiplexed servers, but I suppose a server like rust analyzer could report cargo, clippy and quick diagnostics independently with it.  Anyway, if you count a multiplexer more like plumbing than an actual server, Eglot is still 100% server-agnostic.

**Reply by Askath** [reply]
> > Official in what way? [rassumfrassum](https://github.com/joaotavora/rassumfrassum) is from joaotavora so even if it's not baked in eglot I'd say that's pretty "official".
> 
> It's as "official" as I could make it. I was teasing it/delaying it for years and a few months ago I bit the bullet and started working. It wasn't as hard to do as I thought, to be honest. The main unblocker was using Python, I have to admit.
> 
> > Did you open tickets for the issues you encountered? No matter where or from who the multiplexer is, it's bound to have bugs and kinks at the beginning that need to be ironed out.
> 
> 😃 I second this! 💯
> 
> @nemethf
> 
> > I think a multiplexer could inject its own code-action named "Multiplexer settings/status", which cloud handle "those kinds of things" without the need to modify the LSP client.
> 
> Yes, very much so, as long as these actions are language-agnostic. BTW @nemethf , you'll also notice a very slight break from my usual policy of "Eglot is only for (official) LSP". It now supports a non-standard protocol extension for fast diagnostic interchange where a server can update the diagnostics for a specific aspect of the document and leave other aspects untouched. Obvisouly the "aspects" in use by Rassumfrassum are the underlying multiplexed servers, but I suppose a server like rust analyzer could report cargo, clippy and quick diagnostics independently with it. Anyway, if you count a multiplexer more like plumbing than an actual server, Eglot is still 100% server-agnostic.

I had for lspx, have not tried your solution yet to be fair. Been a few months since I last spend time tinkering with my config just saw the thread being reopened.
Might give it a try next week though!

**Reply by nemethf** [reply]
> > I think a multiplexer could inject its own code-action named "Multiplexer settings/status",

Out of curiosity, I'll try to create a rass PR for this.  (However, I'm not going to tackle backend restarts, because I don't think restarting a single backend server is possible in all cases.  For example, when the restarted server sends capabilities different from its previous answer, the multiplexer cannot amend its own aggregated capabilities to the LSP client.)

>  BTW @nemethf , you'll also notice a very slight break from my usual policy of "Eglot is only for (official) LSP". It now supports a non-standard protocol extension for fast diagnostic interchange ...

Are you talking about `$/streamDiagnostics`?  Is it you who came up with this extension?  If yes, would it make sense to put it under a different "namespace" to avoid a potential name collision?  These are some examples I saw: $ccls/navigate, rust-analyzer/reloadWorkspace, taplo/associatedSchema, ocamllsp/switchImplIntf.  However, not everybody follows this practice: for example, clangd has textDocument/switchSourceHeader.  In any case, it might be a good idea to advertise [the defintion/description](https://github.com/joaotavora/rassumfrassum#protocol-flow) to [the interested parties](https://github.com/microsoft/language-server-protocol/issues).  Or is this only useful for multiplexers?

**Reply by Frestein** [reply]
> A 5-strong mob thinks i should reopen the discussion 😆 ...

Using workarounds is not a solution. See, for example, the implementations of multi-LSP in Neovim and Helix. They don't use any external workarounds on Python. This also applies to the LSP boosters...

**Reply by pbgc** [reply]
> > A 5-strong mob thinks i should reopen the discussion 😆 ...
> 
> Using workarounds is not a solution. See, for example, the implementations of multi-LSP in Neovim and Helix. They don't use any external workarounds on Python. This also applies to the LSP boosters...

I'm (and I'm sure many others) anxiously waiting for your contribution by creating a full multi-LSP for Emacs without workarounds!
While waiting for that I'm VERY thankful for Joao's work both on both eglot and rass.

**Reply by Frestein** [reply]
> > > A 5-strong mob thinks i should reopen the discussion 😆 ...
> > 
> > 
> > Using workarounds is not a solution. See, for example, the implementations of multi-LSP in Neovim and Helix. They don't use any external workarounds on Python. This also applies to the LSP boosters...
> 
> I'm (and I'm sure many others) anxiously waiting for your contribution by creating a full multi-LSP for Emacs without workarounds! While waiting for that I'm VERY thankful for Joao's work both on both eglot and rass.

С чего это я должен делать вклад? Тогда ты говори по русски, лол. 

Твоя попытка перетянуть тему диалога на МЕНЯ с изначальной причины обсуждения, недовольство "эталонным" решением - нелепа. Манипуляции толпой, популизм, коммунистическое "Критикуешь - предлагай!"... Серьезно? Я ожидал от вас большего.

Я высказал вполне себе объективное недовольство решением закрыть дискуссию, потому что не считаю текущее решение - **эталонным**. ИМХО, оно должно быть реализовано на Emacs Lisp и никак иначе. Я открыт к конструктивному диалогу ПОЧЕМУ оно не реализовано и не может быть реализовано **нативно**, как это делают другие.

Я бы донес свою мысль на английском, но если ты не стараешься даже в конструктивный диалог и принятие иных мнений, то и я не буду стремиться изложить информацию на наиболее доступном языке для обсуждения в массе (не потому что я говнюк, просто я слишком плох в английском).

**Reply by joaotavora** [reply]
Yup, when the russian testosterone flares up on this is when I know I should have closed this.  [This 6 year old question was about LSP multiplexers](https://github.com/joaotavora/eglot/discussions/1429#discussion-6976830) .  Now there are at least two.  So if you want to discuss them, post there go post there or on reddit.  If OTOH you want to discuss native multi-server support to Eglot go ahead and open a new discussion here, to suggest how it could be done. Or just fork Eglot.  I'm not gonna implement it, but you may, and if it's good maybe I'll use it!  And don't get maybe don't get so excited, this is all software i give you for free so if you don't like it you don't have to use it.
